webrtc/RTCPeerConnection-onnegotiationneeded.html

```diff
@@ -0,0 +1,287 @@
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.createDataChannel('test');
+    pc.close();
+    return awaitNegotiation(pc);
+
+  }, 'negotiationneeded event should not fire if connection is closed');
+
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.createDataChannel('foo');
+
+    return awaitNegotiation(pc)
+      .then(({nextPromise}) => {
+      pc.createDataChannel('bar');
+      return nextPromise;
+    });
+  }, 'calling createDataChannel twice should fire negotiationneeded event once');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To check if negotiation is needed
+      3.  For each transceiver t in connection's set of transceivers, perform
+          the following checks:
+          1.  If t isn't stopped and isn't yet associated with an m= section
+              according to [JSEP] (section 3.4.1.), return "true".
+
+    5.1.  RTCPeerConnection Interface Extensions
+      addTransceiver
+        9.  Update the negotiation-needed flag for connection.
+   */
+  promise_test(t => {
+    const pc = new RTCPeerConnection();
+    pc.addTransceiver('audio');
+    return awaitNegotiation(pc);
+  }, 'addTransceiver() should fire negotiationneeded event');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To update the negotiation-needed flag
+      4.  If connection's [[needNegotiation]] slot is already true, abort these steps.
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.addTransceiver('audio');
+    return awaitNegotiation(pc)
+    .then(({nextPromise}) => {
+      pc.addTransceiver('video');
+      return nextPromise;
+    });
+  }, 'Calling addTransceiver() twice should fire negotiationneeded event once');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To update the negotiation-needed flag
+      4.  If connection's [[needNegotiation]] slot is already true, abort these steps.
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.createDataChannel('test');
+    return awaitNegotiation(pc)
+    .then(({nextPromise}) => {
+      pc.addTransceiver('video');
+      return nextPromise;
+    });
+  }, 'Calling both addTransceiver() and createDataChannel() should fire negotiationneeded event once');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To update the negotiation-needed flag
+      2.  If connection's signaling state is not "stable", abort these steps.
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    const promise = awaitNegotiation(pc);
+
+    return pc.createOffer()
+    .then(offer => pc.setLocalDescription(offer))
+    .then(() => {
+      assert_equals(pc.signalingState, 'have-local-offer');
+      pc.createDataChannel('test');
+      return promise;
+    });
+  }, 'negotiationneeded event should not fire if signaling state is not stable');
```

Comment:
Agree that the error message from `test_never_resolve` is not very helpful in this case. I was trying to promisify the events so that they can be tested in less imperative way. The `nextPromise` is wrapped in an object so that it doesn't get awaited by the promise resolve algorithm. But I guess using `async_test` with imperative event handlers is probably more readable in this case. Will rewrite the code in new PR for negotiationneeded.